### PR TITLE
Add functionality to use white LEDs on a RGBW strip

### LINKED ISSUE #91: Using white on a SK6812_STRIP_GRBW LED strip
**What do you need help with?**

I have the plugin setup and I would like to use the white LED that comes with the LED strip.

Is this possible?

I have tried to go through the configuration and see if there was an option to enable white but I only see options for RGB

**Log files**
```
[2021-01-30 19:49:23,787] INFO: Initialising LED strip
[2021-01-30 19:49:23,793] ERROR: Strip failed to initialize, no effects will be run.
[2021-01-30 19:49:23,879] ERROR: Please check your settings.
[2021-01-30 19:49:23,913] ERROR: Here's the exception: ws2811_init failed with code -13 (Unable to initialize SPI)
[2021-01-30 19:49:23,945] INFO: No strip initialised, exiting the effect process.
[2021-01-30 19:51:01,452] INFO: Initialising LED strip
[2021-01-30 19:51:01,457] ERROR: Strip failed to initialize, no effects will be run.
[2021-01-30 19:51:01,495] ERROR: Please check your settings.
[2021-01-30 19:51:01,527] ERROR: Here's the exception: ws2811_init failed with code -13 (Unable to initialize SPI)
[2021-01-30 19:51:01,558] INFO: No strip initialised, exiting the effect process.
[2021-01-30 19:51:44,603] INFO: Initialising LED strip
[2021-01-30 19:51:44,609] INFO: Strip successfully initialised
[2021-01-30 19:51:44,609] INFO: Debug logging not enabled, if you are reporting issues please enable it under 'Features' in the settings page.
[2021-01-30 19:51:44,611] INFO: On message recieved, turning on LEDs
[2021-01-30 19:52:49,221] INFO: Off message recieved, turning off LEDs
[2021-01-30 19:52:51,764] INFO: On message recieved, turning on LEDs
[2021-01-30 19:53:44,189] INFO: Off message recieved, turning off LEDs
[2021-01-30 19:53:45,309] INFO: On message recieved, turning on LEDs
[2021-01-30 19:54:01,764] INFO: Kill message recieved, Bye!
[2021-01-30 19:54:01,810] INFO: Initialising LED strip
[2021-01-30 19:54:01,815] ERROR: Strip failed to initialize, no effects will be run.
[2021-01-30 19:54:01,863] ERROR: Please check your settings.
[2021-01-30 19:54:01,905] ERROR: Here's the exception: ws2811_init failed with code -5 (mmap() failed)
[2021-01-30 19:54:01,936] INFO: No strip initialised, exiting the effect process.
[2021-01-30 19:56:46,492] INFO: Initialising LED strip
[2021-01-30 19:56:46,497] INFO: Strip successfully initialised
[2021-01-30 19:56:46,498] INFO: Debug logging not enabled, if you are reporting issues please enable it under 'Features' in the settings page.
[2021-01-30 19:56:46,499] INFO: On message recieved, turning on LEDs
[2021-01-30 19:56:46,499] INFO: Off message recieved, turning off LEDs
```
* [X] I have read the troubleshooting guide found at https://github.com/cp2004/OctoPrint-WS281x_LED_Status/wiki/Troubleshooting-Guide

### PR DESCRIPTION
**Pull request by samwiseg0**
~~I have not tested this yet but here is what I have so far.~~

This version is tested and works. I would like to hide the white brightness bar when the checkbox is not enabled but I could not get that to work? Maybe you can offer some guidance?

Fixes #91

### CODE REVIEW

File: octoprint_ws281x_led_status/templates/settings_strip_modal.jinja2 [hunk 1]
```diff
@@ -16,6 +16,22 @@
                     </select>
                 </div>
             </div>
+            <div class="control-group">
+                <label class="control-label">{{ _('Use white with RGBW') }}</label>
+                <div class="controls">
+                    <input type="checkbox" class="input-medium" data-bind="checked: settings.plugins.ws281x_led_status.white_override">
+                </div>
+            </div>
+            <div class="control-group inline">
```
Comment (cp2004) [inline octoprint_ws281x_led_status/templates/settings_strip_modal.jinja2:25]:
```suggestion
            <div class="control-group inline" data-bind="visible: settingsViewModel.settings.plugins.ws281x_led_status.white_override">
```

That should do it for showing the controls if the setting is enabled

File: octoprint_ws281x_led_status/templates/settings_strip_modal.jinja2 [hunk 2]
```diff
@@ -16,6 +16,22 @@
                     </select>
                 </div>
             </div>
+            <div class="control-group">
+                <label class="control-label">{{ _('Use white with RGBW') }}</label>
+                <div class="controls">
+                    <input type="checkbox" class="input-medium" data-bind="checked: settings.plugins.ws281x_led_status.white_override">
```
Comment (cp2004) [inline octoprint_ws281x_led_status/templates/settings_strip_modal.jinja2:22]: Should this be:
```suggestion
                    <input type="checkbox" class="input-medium" data-bind="checked: settingsViewModel.settings.plugins.ws281x_led_status.white_override">
```
I haven't tested it, but it is the odd one out.

**Comment by cp2004** [comment]
It looks good, that implementation is quite simple but allows for a good amount of usability.

Will test it tomorrow, and merge.

**Comment by samwiseg0** [comment]
@cp2004 I started thinking about the logic a bit more and any compensations that were set would break the the logic for dedicated white. So the solution imo was to process dedicated white _before_ any color correction should be made. This allows full functionality color correction as well as dedicated white. 

The only thing left would be to document that the only way dedicated white will work is if "white" is set to all 255 values and the checkbox it checked. If a user decides to change the "white" value it will not work.

**Review by cp2004 — APPROVED** [review]
👍 
It works for a standard RGB strip, and as expected doesn't show anything if the white setting is enabled on my RGB strip. If I can figure out how the select Knockout bindings work I might be able to make it show conditionally on the type of strip selected, but I'll leave that for now since they can be really annoying. If you want to dive into KO hell, feel free lol.

Thank you for the contribution, I don't think I would have come up with something like this so it is much appreciated 🙂.

**Comment by cp2004** [comment]
Oh, BTW, I have potential first release candidate for 0.7.0 pencilled in to the weekend of the 20/21st of Feb, but knowing previous experience with this plugin I probably shouldn't say that, it will likely slip.

**Comment by samwiseg0** [comment]
> If I can figure out how the select Knockout bindings work I might be able to make it show conditionally on the type of strip selected, but I'll leave that for now since they can be really annoying. If you want to dive into KO hell, feel free lol.

Ha! I was thinking about that as well. Maybe I will take a stab at it! I am back to printing now so I cant test any changes. At lest not for another day or 2. I thought about setting up my other pi just for testing since I still have a bunch of LEDs left. 

> Thank you for the contribution, I don't think I would have come up with something like this so it is much appreciated 🙂.

You are most welcome. I am glad I could contribute in some way to the community. I was very motivated to make this work as I spent more money on those strips just for the functionality!

**Comment by samwiseg0** [comment]
> Oh, BTW, I have potential first release candidate for 0.7.0 pencilled in to the weekend of the 20/21st of Feb, but knowing previous experience with this plugin I probably shouldn't say that, it will likely slip.

Ok sounds great! For now I will be running the latest `devel` as that is the only way I can have my LEDs work. 

I did have a question about the latest build and how it works when the bed is leveling. Is it best to hit you up on discord or would you prefer I open a separate issue to discuss it?

**Comment by cp2004** [comment]
> I did have a question about the latest build and how it works when the bed is leveling. Is it best to hit you up on discord or would you prefer I open a separate issue to discuss it?

I don't mind, if it is a problem then probably an issue, if it is just a quick question discord is fine. There are still problems on the `devel` branch, I am working through a couple now with the fading in/out of effects.

**Pull request merged**